examples/Iris/step3_optimize.cc

```diff
@@ -0,0 +1,435 @@
+  }
+  using ModelVisualizer = NoModelVisualizer;
+#else
+  // A real-life example: assume each of the three classes is defined by a hypersphere-shaped Gaussian.
+  // The five parameters per each class are the centrer { SL, SW, PL, PW } and the raduis of the Gaussian.
+  // The raduis parameter is exponentiated prior to optimizations, to ensure the resulting radius is positive.
+  // Two models are then trained, the second one from the point found by the first one:
+  // Model 1, the descriptive one: Find the best Gaussian per class (could be computed analytically, but why bother?)
+  // Model 2, the discriminant one: Move the best per-class Gaussians to solve the actual classification problem.
+  static std::vector<double> StartingPoint(const std::vector<IrisFlower>& flowers) {
+    // return std::vector<double>(3 * 5, 0.0);
+    // Start with wide Gaussians.
+    std::vector<double> point;
+    for (size_t c = 0; c < 3; ++c) {
+      for (size_t d = 0; d < 4; ++d) {
+        const double IrisFlower::*p = features_list[d].second.mem_ptr;
+        double x = 0.0;
+        size_t n = 0.0;
+        for (const auto& flower : flowers) {
+          const auto label_cit = label_indexes.find(flower.Label);
+          if (label_cit != label_indexes.end() && label_cit->second == c) {
+            x += flower.*p;
+            ++n;
+          }
+        }
+        assert(n > 0);
+        point.push_back(x / n);
+      }
+      point.push_back(-2.0);  // `-log(r ^ 2)` == `-2`, initial `r` = `exp(1)`, a wide Gaussian to start from.
+    }
+    return point;
+  }
+
+  template <typename T>
+  fncas::optimize::ObjectiveFunctionValue<T> ObjectiveFunction(const std::vector<T>& x) const {
+    T penalty_descriptive = 0.0;
+    T penalty_discriminant = 0.0;
+    WeightsComputer<T> computer(x);
+    size_t valid_examples = 0;
+    for (const auto& flower : flowers) {
+      const auto label_cit = label_indexes.find(flower.Label);
+      if (label_cit != label_indexes.end()) {
+        ++valid_examples;
+        T w[3];
+        T ws = 0.0;
+        for (size_t c = 0; c < 3; ++c) {
+          w[c] = computer.WeightInClass(flower, c, x);
+          ws += w[c];
+        }
+        penalty_descriptive += fncas::log(w[label_cit->second]);
+        penalty_discriminant += fncas::log(w[label_cit->second] / ws);
+      }
+    }
+    assert(valid_examples > 0u);
+
+    const T accuracy = fncas::exp(penalty_discriminant / valid_examples);
+
+    return fncas::optimize::ObjectiveFunctionValue<T>([&]() {
+      if (computation_type == ComputationType::ComputeAccuracy) {
+        // Return the probability, and the starting point should be `(1/3)`.
+        return accuracy;
+      } else {
+        // Just return the non-normalized number, for optimization purposes.
+        return computation_type == ComputationType::TrainDescriptiveModel ? penalty_descriptive : penalty_discriminant;
```

Comment:
If you search for the maximum of the objective function, it's definitely not the penalty you want to maximize...